nodego/supervisor.go

```diff
@@ -0,0 +1,315 @@
+}
+
+var loggingCtx loggingContext
+
+var (
+	infoLogWriter  supervisorWriter = "INFO"
+	errorLogWriter supervisorWriter = "ERROR"
+
+	// InfoLogger is a logger that batches sends logs to the supervisor with a
+	// severity level of INFO.
+	InfoLogger = log.New(infoLogWriter, "", 0)
+	// InfoLogger is a logger that batches sends logs to the supervisor with a
+	// severity level of ERROR.
+	ErrorLogger = log.New(errorLogWriter, "", 0)
+)
+
+func init() {
+	if supervisorHostname != "" && supervisorInternalPort != "" {
+		loggingCtx.initialize()
+	}
+}
+
+const isoTimeFormat = "2006-01-02T15:04:05.999Z07:00"
+
+type supervisorWriter string
+
+// Write implements io.Writer.Write.
+func (w supervisorWriter) Write(p []byte) (int, error) {
+	entry := &logEntry{
+		TextPayload: string(p),
+		Severity:    string(w),
+		Time:        time.Now().Format(isoTimeFormat),
+		ExecutionID: loggingCtx.executionID(),
+	}
+
+	if !loggingCtx.addEntry(entry) {
+		return os.Stderr.Write(entry.consoleOutput())
+	}
+
+	return len(entry.TextPayload), nil
+}
+
+func newSupervisorRequest(path string, v interface{}) (*http.Request, error) {
+	postData, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", (&url.URL{
+		Scheme: "http",
+		Host:   supervisorHostname + ":" + supervisorInternalPort,
+		Path:   path,
+	}).String(), bytes.NewBuffer(postData))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Content-Length", strconv.Itoa(len(postData)))
+
+	return req, nil
+}
+
+func doRequestWithContext(ctx context.Context, r *http.Request) (*http.Response, error) {
+	resp, err := http.DefaultClient.Do(r.WithContext(ctx))
+	if err != nil {
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+		default:
+		}
+	}
+	return resp, err
+}
+
+func postToSupervisor(path string, v interface{}, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	req, err := newSupervisorRequest(path, v)
+	if err != nil {
+		return err
+	}
+
+	resp, err := doRequestWithContext(ctx, req)
+	if err == ctx.Err() {
+		return errors.New("timeout when calling supervisor")
+	} else if err != nil {
+		return fmt.Errorf("error when calling supervisor: %s\n\n%s\n", err.Error(), debug.Stack())
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		err = fmt.Errorf("incorrect response code from supervisor: %d\n", resp.StatusCode)
+	}
+
+	return err
+}
+
+func killInstance() {
+	err := postToSupervisor("/_ah/kill", nil, supervisorKillTimeout)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+	}
+	// Exit code 16 is copied over from worker.js
+	os.Exit(16)
+}
+
+func loggerMiddleware(handler http.Handler) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		loggingCtx.setExecutionID(r.Header.Get("Function-Execution-Id"))
+
+		defer func() {
+			loggingCtx.setExecutionID("")
+			go loggingCtx.flush()
```

Comment:
Ok, The current synchronization seems correct, although it could probably be simplified. That does not need to happen in this PR though.

* I agree that the current flush can be eliminated. It does not seem to serve a useful purpose.
* I agree that starting new log batches is not needed to ensure that logs are flushed in a timely manor.

That said, I would still like to see new log batches created in the middleware. Here is why:
The new execution ID is only picked up when starting a log batch. As it currently stands, if there is a currently going log batch from before the handler starts, it could get either no execution ID attached or even the wrong execution ID attached.

Reply:
The log batches themselves has no reference to any execution ID. Execution IDs are attached to the log entry when they are created.

That being said, I do understand wanting to separate batches by execution.

Reply:
Oh, good point. Well, in that case I guess we can just remove the flush.